cde/doc/cde1/man/man1/dtscreen.1
.\"--- 
.\"    (c) Copyright 1993, 1994 Hewlett-Packard Company
.\"    (c) Copyright 1993, 1994 International Business Machines Corp.
.\"    (c) Copyright 1993, 1994 Sun Microsystems, Inc.
.\"    (c) Copyright 1993, 1994 Novell, Inc.
.\"--- 
.TH dtscreen 1X "17 Feb. 1994"
.BH "17 Feb. 1994""
.SH NAME
dtscreen - the Common Desktop Environment animated screen savers
.SH SYNOPSIS
dtscreen [
.BI \-display " dsp"
] [
.BI \-delay " usecs"
]
.P
 [
.BI \-batchcount " num"
] [
.BI \-saturation " value"
]
.P
 [
.BI \-nice " nicelevel"
] [
.BI \-mode " mode"
]
.SH DESCRIPTION
The dtscreen client supports the following key tasks.
.RS
.P
- Draws an animated image on a provided window id(s).
.P
- Does not assume the size of the provided window(s). 
.P
- Provides a set of selectable animated images.
.P
- Animation parameters such as nice, speed, number and saturation may be specified.
.RE
.PP
The dtscreen client provides the default screen saver animations for
the desktop. When launched, it will query the desktop using the
.BDtSaverGetWindows()
API for the set for window ids on which to
draw. These window ids could be the cover windows created by the
session manager for session lock, or a window imbedded in the style
manager's session lock customization dialog.  Once obtained, the
dtscreen client will proceed to draw in these windows using the
specified options. If dtscreen cannot obtain the set of window ids, it
will create its own. Normally, the dtscreen client will be invoked
directly by the session manager, or by the style manager.
.P
Note that the Session Manager, dtsession, is responsible for locking
the session and prompting for a password to unlock.
.P
The session manager may launch the dtscreen client to provide screen
saver animations during session lock. Refer to the session manager
specification for resources controlling the launching of screen savers
by the session manager.
.P
The style manager may launch the dtscreen client to provide a preview
of a screen saver animation during customization. Refer to the style
manager specification for information on how to integrate a screen
saver client such as dtscreen into the desktop.
.P
The dtsession client provides a selectable set of animations. These are:
.IP \f3hop\fP 1.5i
Hopalong iterated fractals.
.IP \f3life\fP 1.5i
Conway's game of life.
.IP \f3qix\fP 1.5i
Spinning lines.
.IP \f3image\fP 1.5i
Random bouncing image.
.IP \f3swarm\fP 1.5i
Swarm of bees.
.IP \f3rotor\fP 1.5i
Rotor
.IP \f3pyro\fP 1.5i
Fireworks.
.IP \f3flame\fP 1.5i
Cosmic Flame Fractals.
.IP \f3worm\fP 1.5i
Wiggly Worms.
.IP \f3random\fP 1.5i
Random mode.
.P
The dtscreen client animations are based on xlock(1).
.SH OPTIONS
.sp .5
.BI \-display " dsp"
.RS
The display on which animations will be drawn.
.RE
.sp .5
.BI \-delay " usecs"
.RS
The delay option sets the speed at which a mode will operate. It
simply sets the number of microseconds to delay between batches of
"hopalong pixels", "qix lines", "life generations", "image blits", and
"swarm motions".
.RE
.sp .5
.BI \-batchcount " num"
.RS
The batchcount option sets the number of things to do per batch to
num. In qix mode this refers to the number of lines rendered in the
same color. In hop mode this refers to the number of pixels rendered
in the same color. In image mode this refers to the number of X logos
on the screen at once. in swarm mode this refers to the number of
bees. In life mode it means nothing.
.RE
.sp .5
.BI \-saturation " value"
.RS
The saturation option sets the saturation of the color ramp used to
value. 0 is grayscale, 1 is very rich color and 0.4 is a nice pastel.
.RE
.sp .5
.BI \-nice " nicelevel"
.RS
The nice option sets the system nicelevel of the dtscreen process to
nicelevel.
.RE
.sp .5
.BI \-mode " mode"
.RS
The mode option specifies which animation should be displayed. Values
are:
.RE
.IP "\f3hop\fP" 1.5i
Hopalong iterated fractals.
.IP "\f3life\fP" 1.5i
Conway's game of life.
.IP "\f3qix\fP" 1.5i
 Spinning lines.
.IP "\f3image\fP" 1.5i
 Random bouncing image.
.IP "\f3swarm\fP" 1.5i
 Swarm of bees.
.IP "\f3rotor\fP"  1.5i
 Rotor
.IP "\f3pyro\fP" 1.5i
 Fireworks.
.IP "\f3flame\fP" 1.5i
 Cosmic Flame Fractals.
.IP "\f3worm\fP" 1.5i
 Wiggly Worms.
.IP "\f3random\fP" 1.5i
 Random mode.
.SH RETURN VALUE
 Exit values are:
.IP "\f30\fP" 1.5i
Successful completion.
.IP "\f3>0\fP" 1.5i
Error condition occurred.
.SH EXAMPLES
.sp .5
.SS dtscreen -mode swarm
Display the swarm animation.
.SH RESOURCES
.TS
center;
cf3 cf3 cf3 cf3
l l l l .
Name	ClassClass	Type 	Default
_
mode	Mode	String	swarm
nice	Nice	Int	16
delay	Delay	Int	1000
batchcount	Batchcount	Int	100
saturation	Saturation	Float	1.0
hop.delay	Delay	Int	0
hop.batchcount	Batchcount	Int	1000
hop.saturation	Saturation	Float	1
image.delay	Delay	Int	2000000
image.batchcount	Batchcount	Int	8
image.saturation	Saturation	Float	0.2
qix.delay	Delay	Int	30000
qix.batchcount	Batchcount	Int	64
qix.saturation	Saturation	Float	1
life.delay	Delay	Int	1000000
life.batchcount	Batchcount	Int	1
life.saturation	Saturation	Float	1
swarm.delay	Delay	Int	10000
swarm.batchcount	Batchcount	Int	100
swarm.saturation	Saturation	Float	1
rotor.delay	Delay	Int	10000
rotor.batchcount	Batchcount	Int	4
rotor.saturation	Saturation	Float	0.4
pyro.delay	Delay	Int	15000
pyro.batchcount	Batchcount	Int	40
pyro.saturation	Saturation	Float	1.0
flame.delay	Delay	Int	10000
flame.batchcount	Batchcount	Int	20
flame.saturation	Saturation	Float	1.0
worm.delay	Delay	Int	10000
worm.batchcount	Batchcount	Int	20
worm.saturation	Saturation	Float	1.0
.TE
.SS delay
The delay option sets the speed at which a mode will operate. It
simply sets the number of microseconds to delay between batches of
"hopalong pixels", "qix lines", "life generations", "image blits", and
"swarm motions".
.sp .5
.SS batchcount
The batchcount option sets the number of things to do per batch to
num. In qix mode this refers to the number of lines rendered in the
same color. In hop mode this refers to the number of pixels rendered
in the same color. In image mode this refers to the number of X logos
on the screen at once. in swarm mode this refers to the number of
bees. In life mode it means nothing.
.sp .5
.SS saturation
The saturation option sets the saturation of the color ramp used to
value. 0 is grayscale, 1 is very rich color and 0.4 is a nice pastel.
.sp .5
.SS nice
The nice option sets the system nicelevel of the dtscreen process to
nicelevel.
.sp .5
.SS mode
The mode option specifies which animation should be displayed. Values
are:
.IP hop 1.5i
 Hopalong iterated fractals.
.IP life 1.5i
 Conway's game of life.
.IP qix 1.5i
 Spinning lines.
.IP image 1.5i
 Random bouncing image.
.IP swarm 1.5i
 Swarm of bees.
.IP rotor 1.5i
 Rotor
.IP pyro 1.5i
 Fireworks.
.IP flame 1.5i
 Cosmic Flame Fractals.
.IP worm 1.5i
 Wiggly Worms.
.IP random 1.5i
 Random mode.
.SH SEE ALSO
.IP dtsession 1.5i
Launches the dtscreen client during session lock.
Provides resources that can be used to control this event. Locks and
unlocks a session.
.IP dtstyle 1.5i
Launched the dtscreen client during session lock
customization. Provides the methods by which additional screen saver
clients may be integrated into the desktop.


cde/cde1/dtscreen/dtscreen.c
/* $XConsortium: dtscreen.c /main/cde1_maint/2 1995/08/29 19:57:04 gtsang $ */
/*
 */

/*-
 * dtscreen.c - X11 client to lock a display and show a screen saver.
 *
 * Copyright (c) 1988-91 by Patrick J. Naughton.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * This file is provided AS IS with no warranties of any kind.  The author
 * shall have no liability with respect to the infringement of copyrights,
 * trade secrets or any patents by this file or any part thereof.  In no
 * event will the author be liable for any lost revenue or profits or
 * other special, indirect and consequential damages.
 *
 * Comments and additions should be sent to the author:
 *
 *                     naughton@eng.sun.com
 *
 *                     Patrick J. Naughton
 *                     MS 21-14
 *                     Sun Laboritories, Inc.
 *                     2550 Garcia Ave
 *                     Mountain View, CA  94043
 *
 * Revision History:
 * 16-May-91: added pyro and random modes.
 *            ripped big comment block out of all other files.
 * 29-Oct-90: added cast to XFree() arg.
 *            added volume arg to call to XBell().
 * 31-Aug-90: added blank mode.
 *            added swarm mode.
 *            moved usleep() and seconds() out to usleep.c.
 *            added SVR4 defines to dtscreen.h
 * 29-Jul-90: added support for multiple screens to be locked by one dtscreen.
 *            moved global defines to dtscreen.h
 *            removed use of allowsig().
 * 07-Jul-90: reworked commandline args and resources to use Xrm.
 *            moved resource processing out to resource.c
 * 02-Jul-90: reworked colors to not use dynamic colormap.
 * 23-May-90: added autoraise when obscured.
 * 18-Feb-90: added SunOS3.5 fix.
 *            changed -mono -> -color, and -saver -> -lock.
 *            allow non-locking screensavers to display on remote machine.
 *            added -echokeys to disable echoing of '?'s on input.
 *            cleaned up all of the parameters and defaults.
 * 23-Sep-89: Added fix to allow local hostname:0 as a display.
 *            Put empty case for Enter/Leave events.
 *            Moved colormap installation later in startup.
 * 19-Sep-89: Added '?'s on input.
 * 27-Mar-89: Added -qix mode.
 *            Fixed GContext->GC.
 * 08-Mar-89: Added -nice, -mode and -display, built vector for life and hop.
 * 24-Feb-89: Replaced hopalong display with life display from SunView1.
 * 22-Feb-89: Added fix for color servers with n < 8 planes.
 * 16-Feb-89: Updated calling conventions for XCreateHsbColormap();
 *            Added -count for number of iterations per color.
 *            Fixed defaulting mechanism.
 *            Ripped out VMS hacks.
 *            Sent to expo for X11R3 contrib.
 * 19-Jan-89: Fixed monochrome gc bug.
 * 19-Sep-88: Changed -color to -mono. (default is color on color displays).
 *            Added -saver option. (just do display... don't lock.)
 * 01-Apr-88: Added XGrabServer/XUngrabServer for more security.
 * 27-Mar-88: Rotate fractal by 45 degrees clockwise.
 * 24-Mar-88: Added color support. [-color]
 *            wrote the man page.
 * 23-Mar-88: Added HOPALONG routines from Scientific American Sept. 86 p. 14.
 *            added requirement for display to be "unix:0".
 * 22-Mar-88: Recieved Walter Milliken's comp.windows.x posting.
 *
 */

/*              include file for message texts          */
#include <limits.h>
#include <nl_types.h>
#define MF_DTSCREEN "dtscreen"

#include <nl_types.h>
#include <locale.h>
nl_catd  scmc_catd;   /* Cat descriptor for scmc conversion */

#include <stdio.h>
#include <signal.h>
#include <string.h>

#include <X11/Intrinsic.h> /* For Boolean */
#include "dtscreen.h"
#include <Dt/EnvControl.h>
#include <Dt/Saver.h>

#if !defined(NL_CAT_LOCALE)
#define NL_CAT_LOCALE 0
#endif

char       *ProgramName;        /* argv[0] */
Display    *dsp = NULL;         /* server display connection */
void        (*callback) () = NULL;
void        (*init) () = NULL;

static perwindow *Win; /* perwindow information */
static int  windows;            /* number of windows */
static Window *winprop = NULL;  /* dtsession cover windows */

/* VARARGS1 */
void
error(s1, s2)
    char       *s1, *s2;
{
    fprintf(stderr, s1, ProgramName, s2);
    exit(1);
}

/*
 * Restore all grabs, reset screensaver, restore colormap, close connection.
 */
static void
finish()
{
    XSync(dsp, False);
    if (winprop) 
      XFree((char *)winprop);
    XFlush(dsp);
    XCloseDisplay(dsp);
}


static void
justDisplay()
{
    XEvent      event;
    int window;

    for (window = 0; window < windows; window++)
        init(Win+window);
    do {
        while (!XPending(dsp)) {
            for (window = 0; window < windows; window++)
                callback(Win+window);
            XSync(dsp, False);
            usleep(delay);
        }
        XNextEvent(dsp, &event);

#ifndef DEBUG
        if (event.type == VisibilityNotify)
            XRaiseWindow(dsp, event.xany.window);
#endif
    } while (1);
}


long
allocpixel(cmap, name, def)
    Colormap    cmap;
    char       *name;
    char       *def;
{
    XColor      col;
    XColor      tmp;
    XParseColor(dsp, cmap, name, &col);
    if (!XAllocColor(dsp, cmap, &col)) {
#ifdef MIT_R5

        fprintf(stderr, 
             "couldn't allocate: %s, using %s instead\n",
                name, def);

#else
        fprintf(stderr, catgets(scmc_catd, 2, 35, 
             "couldn't allocate: %s, using %s instead\n"),
                name, def);
#endif

        XAllocNamedColor(dsp, cmap, def, &col, &tmp);
    }
    return col.pixel;
}


int
main(argc, argv)
    int         argc;
    char       *argv[];
{
    XSetWindowAttributes xswa;
    XGCValues   xgcv;
    XColor      nullcolor;
    int         nitems = 0;
    int         window;
    int i;

   /*
    * So we can find the app-defaults file.
    */
    _DtEnvControl(DT_ENV_SET);

    /* set locale to what is defined by local environment */
    setlocale(LC_ALL,"");
    scmc_catd = catopen(MF_DTSCREEN, NL_CAT_LOCALE);

    ProgramName = strrchr(argv[0], '/');
    if (ProgramName)
        ProgramName++;
    else
        ProgramName = argv[0];

    srandom(time((time_t *) 0));  /* random mode needs the seed set. */

    GetResources(argc, argv);

    CheckResources();

   if (!create)
   {
     /*
      * Try to get array of window ids from desktop. Also returned is
      * a property identifier whose change will tell us when to terminate.
      */
      if (!DtSaverGetWindows(dsp, &winprop, &nitems))
      {
        fprintf(stderr, catgets(scmc_catd, 2, 40,
             "%s: Cannot locate window in which to draw. Using the -create\n"
             "option will cause %s to create its own window.\n"),
             ProgramName, ProgramName);

        exit(1);
      }

      Win = (perwindow *)malloc(nitems * sizeof(perwindow));
      windows = nitems;
      for (window = 0; window < windows; window++)
      {
        Win[window].w = winprop[window];
        Win[window].perscreen = NULL;
        Win[window].data = NULL;
      }
    }
    else
    {
      int screen_num = DefaultScreen(dsp);
      Screen     *scr = ScreenOfDisplay(dsp, screen_num);


      Win = (perwindow *)malloc(sizeof(perwindow));
      windows = 1;
      xswa.override_redirect = True;
      xswa.background_pixel = BlackPixelOfScreen(scr);
      xswa.event_mask = VisibilityChangeMask;

      Win[0].w = XCreateWindow(
                      dsp,
                      RootWindowOfScreen(scr),
                      0, 0,
                      WidthOfScreen(scr) - 100, HeightOfScreen(scr) - 100,
                      0, CopyFromParent, InputOutput, CopyFromParent,
                      CWBackPixel | CWEventMask, &xswa);
      Win[0].perscreen = NULL;
      Win[0].data = NULL;
      XMapWindow(dsp, Win[0].w);
      XRaiseWindow(dsp, Win[0].w);
    }

    for (window = 0; window < windows; window++)
    {
      XWindowAttributes attr;

      XGetWindowAttributes(dsp, Win[window].w, &attr);

      for (i = 0; i < window; i++)
      {
        if (Win[i].perscreen && Win[i].perscreen->screen == attr.screen)
        {
          Win[window].perscreen = Win[i].perscreen;
          break;
        }
      }

      if (Win[window].perscreen == NULL)
      {
        Colormap cmap;
        perscreen *pscreen;

        pscreen = (perscreen *)malloc(sizeof(perscreen));
        Win[window].perscreen = pscreen;

        pscreen->screen = attr.screen;
        cmap = DefaultColormapOfScreen(pscreen->screen);
        
        pscreen->pixels[0] = allocpixel(cmap, "White", "White");
        pscreen->pixels[1] = allocpixel(cmap, "Black", "Black");
        if (mono || CellsOfScreen(pscreen->screen) == 2)
        {
          pscreen->npixels = 2;
        }
        else
        {
          int         colorcount = NUMCOLORS;
          u_char      red[NUMCOLORS];
          u_char      green[NUMCOLORS];
          u_char      blue[NUMCOLORS];
  
          hsbramp(0.0, saturation, 1.0, 1.0, saturation, 1.0, colorcount,
                      red, green, blue);
          pscreen->npixels = 0;
          for (i = 0; i < colorcount; i++)
          {
            XColor      xcolor;

            xcolor.red = red[i] << 8;
            xcolor.green = green[i] << 8;
            xcolor.blue = blue[i] << 8;
            xcolor.flags = DoRed | DoGreen | DoBlue;

            if (!XAllocColor(dsp, cmap, &xcolor))
                break;

            pscreen->pixels[i] = xcolor.pixel;
            pscreen->npixels++;
          }
        }
      }

      xgcv.foreground = WhitePixelOfScreen(Win[window].perscreen->screen);
      xgcv.background = BlackPixelOfScreen(Win[window].perscreen->screen);
      Win[window].gc = XCreateGC(dsp, Win[window].w,
                         GCForeground | GCBackground, &xgcv);
    }
    nice(nicelevel);

    justDisplay();

    finish();

    return 0;
}
